### DISCUSSION TITLE
The inferred type of 'auth' cannot be named without a reference

### DISCUSSION THREAD

**Discussion by yordis** [nextauthjs/next-auth#9950] [Help] [answered]
hey folks, [I am trying to upgrade the setup to V5](https://github.com/straw-hat-team/nextjs-starter-kit/pull/157) with a very basic auth file (I commented the code to test it out), TypeScript compiler is giving me some trouble:

```ts
import NextAuth, { NextAuthConfig } from 'next-auth';
const config = {

} satisfies NextAuthConfig;
export const { handlers, auth, signIn, signOut } = NextAuth(config);
```

```
./src/auth.ts:67:26
Type error: The inferred type of 'auth' cannot be named without a reference to '../../../.yarn/__virtual__/next-auth-virtual-b4c85b046e/0/cache/next-auth-npm-5.0.0-beta.8-b5debcb93d-4458f65a76.zip/node_modules/next-auth/lib'. This is likely not portable. A type annotation is necessary.
```

**Comment by yordis** [comment]
Hey @ThangHuuVu, I sincerely apologize for tagging you directly. I am unsure what to do anymore. Have you seen this issue before or know what the issue is with TypeScript?

PR: https://github.com/straw-hat-team/nextjs-starter-kit/pull/157

**Reply by ThangHuuVu** [reply]
It seems to be a TS issue with your yarn setup - Sorry I'm not able to help further in this case, it's not an issue with `next-auth`. Maybe you can try [this](https://github.com/microsoft/TypeScript/issues/29808#issuecomment-540292885) (found it after a quick GH issue search)

**Reply by yordis** [reply]
I removed `yarn pnp`, and the problem still exists 🤔 

```
./src/auth.ts:41:26
Type error: The inferred type of 'auth' cannot be named without a reference to '../../../node_modules/next-auth/lib'. This is likely not portable. A type annotation is necessary.
```

**Reply by yordis** [reply]
I see the following https://github.com/nextauthjs/next-auth/blob/e179d0b94c08b014a37701d7a8a0f5d0add08c21/packages/next-auth/src/index.tsx#L398

Which ring the bell if maybe that is the issue 🤔 why is it expecting an error?

**Reply by yordis** [reply]
Hey @ThangHuuVu, I removed `yarn` altogether and used `pnpm`, and the issue still persists...

![Screenshot 2024-02-13 at 12 21 04 AM](https://github.com/nextauthjs/next-auth/assets/4237280/e3be03b5-4c0d-4b6f-a3b7-c5ade03f88c0)

**Reply by xentobias** [reply]
I have the same problem and I'm using npm. Next-Auth (5.0.0-beta.13)

**Reply by yordis** [reply]
Should I create an issue thou? At this point I do not believe it is me then

**Reply by nelsonlaidev** [reply]
![CleanShot 2024-02-27 at 21 49 29@2x](https://github.com/nextauthjs/next-auth/assets/75498339/c9e2aadd-b349-4c26-911a-dc1948aca6af)

I got the same problem with `5.0.0-beta.13`.

**Reply by xentobias** [reply]
@yordis, it would probably be really helpful if you could create an issue.

A (not so elegant) temporary workaround would be:
```typescript
const nextAuthResult = NextAuth({
  // ...
})

export const auth: NextAuthResult["auth"] = nextAuthResult.auth;
export const {
  handlers: { GET, POST },
  signIn,
  signOut,
}: NextAuthResult = nextAuthResult;
```

**Comment by yordis** [comment]
After trying multiple things following https://github.com/microsoft/TypeScript/issues/42873 I decided to add to my Next Apps computation:

```json
{
  "compilerOptions": {
    "declaration": false,
  }
}
```

It is fixed under `"@straw-hat/tsconfig": "4.2.2"`

I am not sure what the downsides are doing that, `for app compilations`, I would love to hear your opinions here:

**Reply by Falven** [reply]
How is disabling declarations the accepted answer???

**Reply by yordis** [reply]
Since initiating this discussion, I determined my best solution and answer. I know how much time somebody could spend on the topic since I have already spent many hours on it. I tried everything people suggested or combination of things.

I'm open to hearing if you have a more practical approach; I am willing to turn off an app's declaration, but I wish I did not have to be that way; as you noticed, I asked for opinions about it.

That said, please contribute a constructive alternative if you have one. Otherwise, your feedback comes across as unproductive criticism.

**Reply by Falven** [reply]
It's productive criticism. This shouldn't be accepted as an answer as it's not the correct answer for apps with a well defined public interface.

**Reply by yordis** [reply]
What is the proper solution? Could you focus on that? I am more than happy to mark your solution as the answer.

Otherwise,

![image](https://github.com/user-attachments/assets/eb5b2c0e-8e17-4b87-894c-b476063c0d45)

**Reply by Falven** [reply]
One option is to explicitly re-type the destructed declarations.

```ts
import authOptions from './config/auth.config';
import NextAuth, { type NextAuthResult } from 'next-auth';

const result = NextAuth(authOptions);

export const handlers: NextAuthResult['handlers'] = result.handlers;
export const auth: NextAuthResult['auth'] = result.auth;
export const signIn: NextAuthResult['signIn'] = result.signIn;
export const signOut: NextAuthResult['signOut'] = result.signOut;
```

This avoids the naming issue because each property is typed individually and doesn't require a single stable type name for the entire object. But it's really up to `next-auth` to make their declarations more stable.

**Reply by chrislgrigg** [reply]
Beautiful, @Falven, thank you.

**Reply by MukeshAbhi** [reply]
ThankYou @Falven

**Comment by LouisLecouturier** [comment]
I am having the same issue but with the `signIn` variable

```
TS2742: The inferred type of signIn cannot be named without a reference to .pnpm/@auth+core@0.31.0/node_modules/@auth/core/providers. This is likely not portable. A type annotation is necessary.
```


My code :
```ts
import NextAuth from 'next-auth';
import Entra from 'next-auth/providers/microsoft-entra-id';

const {
  AZURE_ENTRA_TENANT_ID,
  AZURE_ENTRA_CLIENT_ID,
  AZURE_ENTRA_CLIENT_SECRET,
} = process.env;


export const { auth, handlers, signIn, signOut } : NextAuthResult = NextAuth({ // The signIn is marked as error
  providers: [
    Entra({
      tenantId: AZURE_ENTRA_TENANT_ID,
      clientId: AZURE_ENTRA_CLIENT_ID,
      clientSecret: AZURE_ENTRA_CLIENT_SECRET,
    }),
  ],
});
```

**Reply by olivierlesnicki** [reply]
Likewise

**Reply by theSaiRohit** [reply]
Any updates?

**Comment by mrblackoutz** [comment]
Actually I don't know why this is closed, the problem persists, trying to fix the Profiles I got to:
```ts
import NextAuth, { NextAuthConfig, NextAuthResult } from "next-auth";
import GoogleProvider, { GoogleProfile } from "next-auth/providers/google";
import MicrosoftEntraProvider, { MicrosoftEntraIDProfile } from "next-auth/providers/microsoft-entra-id";

// Define explicit types for the NextAuth configuration
const authOptions: NextAuthConfig = {
 providers: [
 GoogleProvider({
 authorization: {
 params: {
 prompt: "consent",
 access_type: "offline",
 response_type: "code",
 },
 },
 }),
 MicrosoftEntraProvider({
 clientId: process.env.AUTH_MICROSOFT_ENTRA_ID_ID || '',
 clientSecret: process.env.AUTH_MICROSOFT_ENTRA_ID_SECRET || '',
 issuer: `https://login.microsoftonline.com/${process.env.AUTH_MICROSOFT_ENTRA_ID_TENANT || 'common'}/v2.0`,
 authorization: {
 params: {
 scope: 'openid profile email User.Read',
 prompt: 'select_account'
 }
 }
 })
 ],
 session: { strategy: "jwt" },
 pages: {
 signIn: "/login",
 }
};

// Infer the return type of NextAuth
type NextAuthReturnType = NextAuthResult;

type Profiles = GoogleProfile | MicrosoftEntraIDProfile;

// Define custom types for signIn and signOut with correct generics
type CustomSignIn = <P extends Profiles, R extends boolean = true>(
 provider?: P,
 options?:
 | FormData
 | ({
 redirectTo?: string
 redirect?: R
 } & Record<string, any>),
 authorizationParams?:
 | string[][]
 | Record<string, string>
 | string
 | URLSearchParams
) => Promise<R extends false ? any : never>;
type CustomSignOut = NextAuthReturnType['signOut'];

// Export the handlers explicitly with custom types
export const { handlers, signIn, signOut, auth } = NextAuth(authOptions) as {
handlers: NextAuthReturnType['handlers'];
signIn: CustomSignIn;
signOut: CustomSignOut;
auth: NextAuthReturnType['auth'];
};
```

but, it definitely doesn't seem like a reasonable solution to me, @Falven's solution worked for me, I'll copy the error in my native language, so that other people can also find this solution
`O tipo inferido de 'signIn' não pode ser nomeado sem uma referência a '.pnpm/@auth+core@0.39.0/node_modules/@auth/core/providers'. Isso provavelmente não é portátil. Uma anotação de tipo é necessária. ts(2742)`

The solution proposed by @Falven and it worked for me was:
```ts
const result = NextAuth(authOptions);

export const handlers: NextAuthResult['handlers'] = result.handlers;
export const auth: NextAuthResult['auth'] = result.auth;
export const signIn: NextAuthResult['signIn'] = result.signIn;
export const signOut: NextAuthResult['signOut'] = result.signOut;
```

I found the issue for this here #10568

**Comment by TomBell95** [comment]
I believe this is related to the setup in `tsconfig.json` - I just tried to import an existing (working) next app with auth into my turborepo and found the below **does not work**:

```json
// tsconfig.json
{
  "extends": "@repo/typescript-config/nextjs.json",
  "compilerOptions": {
    "plugins": [
      {
        "name": "next"
      }
    ],
    "paths": {
      "@/*": ["./src/*"]
    }
  },
  "include": ["**/*.ts", "**/*.tsx", "next-env.d.ts", "next.config.js", ".next/types/**/*.ts"],
  "exclude": ["node_modules"]
}

// typescript-config/nextjs.json
{
  "$schema": "https://json.schemastore.org/tsconfig",
  "extends": "./base.json",
  "compilerOptions": {
    "plugins": [{ "name": "next" }],
    "module": "ESNext",
    "moduleResolution": "Bundler",
    "allowJs": true,
    "jsx": "preserve",
    "noEmit": true
  }
}

```


Whilst the original **does work**:

```json
// working tsconfig.json
{
  "compilerOptions": {
    "target": "ES2017",
    "lib": ["dom", "dom.iterable", "esnext"],
    "allowJs": true,
    "skipLibCheck": true,
    "strict": true,
    "noEmit": true,
    "esModuleInterop": true,
    "module": "esnext",
    "moduleResolution": "bundler",
    "resolveJsonModule": true,
    "isolatedModules": true,
    "jsx": "preserve",
    "incremental": true,
    "plugins": [
      {
        "name": "next"
      }
    ],
    "paths": {
      "@/*": ["./src/*"]
    }
  },
  "include": ["next-env.d.ts", "**/*.ts", "**/*.tsx", ".next/types/**/*.ts"],
  "exclude": ["node_modules"]
}

```

**Reply by DrewLandgraveCbsi** [reply]
This works for me as well. Not the correct solution but interesting that the original is working

**Reply by Codec-Bro8** [reply]
haha, this solution works. Cant believe this issue isn't fixed yet from next-auth

**Comment by yordis** [comment, marked as answer]
From @Falven, check https://github.com/nextauthjs/next-auth/discussions/9950#discussioncomment-11546915

One option is to explicitly re-type the destructed declarations.

```ts
import authOptions from './config/auth.config';
import NextAuth, { type NextAuthResult } from 'next-auth';

const result = NextAuth(authOptions);

export const handlers: NextAuthResult['handlers'] = result.handlers;
export const auth: NextAuthResult['auth'] = result.auth;
export const signIn: NextAuthResult['signIn'] = result.signIn;
export const signOut: NextAuthResult['signOut'] = result.signOut;
```

This avoids the naming issue because each property is typed individually and doesn't require a single stable type name for the entire object. But it's really up to next-auth to make their declarations more stable.

**Reply by johnrommeloctaviano** [reply]
Works for me. Thank you @yordis!

**Comment by ValentinGurkov** [comment]
Thanks for providing a workaround. Explicitly re-typing the declarations is a bit ugly though, I hope this gets addressed.

**Comment by DrewLandgraveCbsi** [comment]
Ok coming back to this after a few months because I ran into something similar with react types. I had some turbo generators that were pinning the @types/react package to 19.0.8 and other apps in my turborepo using pnpm had @types/react at 19.1.8. The mismatch in versions of these packages caused the issue. 

For me I'm able to update all the versions to the same version and just have a caret there now so it should bring in the latest minor/patch versions.
